src/Compilers/CSharp/Portable/CodeGen/CodeGenerator.cs

```diff
@@ -401,6 +407,58 @@ private void EmitSequencePointStatement(BoundSequencePointWithSpan node)
             }
         }
 
+        private void EmitSaveSequencePoint(BoundSaveSequencePoint statement)
+        {
+            if (!_emitPdbSequencePoints)
+                return;
+
+            ArrayBuilder<RawSequencePoint> sequencePoints = _builder.SeqPointsOpt;
+            if (sequencePoints is null)
+                return;
+
+            for (int i = sequencePoints.Count - 1; i >= 0; i--)
+            {
+                var span = sequencePoints[i].Span;
+                if (span == RawSequencePoint.HiddenSequencePointSpan)
+                    continue;
+
+                // Found the previous non-hidden sequence point.  Save it.
+                _savedSequencePoints ??= PooledDictionary<object, TextSpan>.GetInstance();
+                _savedSequencePoints.Add(statement.Identifier, span);
+                return;
+            }
+        }
+
+        private void EmitRestorePreviousSequencePoint(BoundRestorePreviousSequencePoint node)
+        {
+            Debug.Assert(node.Syntax is { });
+            if (_savedSequencePoints is null || !_savedSequencePoints.TryGetValue(node.Identifier, out var span))
+                return;
+
+            EmitStepThroughSequencePoint(node.Syntax.SyntaxTree, span);
+        }
+
+        private void EmitStepThroughSequencePoint(BoundStepThroughSequencePoint node)
+        {
+            EmitStepThroughSequencePoint(node.Syntax.SyntaxTree, node.Span);
+        }
+
+        private void EmitStepThroughSequencePoint(SyntaxTree syntaxTree, TextSpan span)
+        {
+            if (!_emitPdbSequencePoints)
+                return;
+
+            var label = new object();
+            // The IL builder is eager to discard unreachable code, so
+            // we fool it by branching on a condition that is always true at runtime.
+            _builder.EmitConstantValue(ConstantValue.Create(true));
+            _builder.EmitBranch(ILOpCode.Brtrue, label);
+            EmitSequencePoint(syntaxTree, span);
+            _builder.EmitOpCode(ILOpCode.Nop);
```

Comment:
@gafter Just curious if sequence points can be empty w/o instructions, why having EmitSequencePoint() followed by EmitHiddenSequencePoint() is not sufficient?

I'm asking because coverage tools show uncovered statements, for example: https://developercommunity.visualstudio.com/content/problem/1140065/coverage-on-switch-expression-dropped-in-167.html , and now they should always ignore the unreachable SP in `ldc.i4.1` `brtrue.s 1` `nop` pattern. Is there a guarantee the user's code will never produce such IL?

Reply:
The user's code can indeed produce such IL:
https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1RhYAdsAQLIAUAlBVTTwG4CGAJwJMAvCMEBXAKYBubj2pgAZgVbBO87IprIA7AXRaeAXxwmgA=